odl/deform/linear_deformation.py

```diff
@@ -0,0 +1,192 @@
+    def __init__(self, offsets, vector_field_jacobian, divergence=None):
+        """Initialize a linear deformation
+
+        Parameters
+        ----------
+        vector_field : `ProductSpace` of `DiscreteLp` spaces
+            The (small) deformation field.
+        vector_field_jacobian : `DiscreteLp`
+            The determinant of the jacobian of the deformation.
+        divergence : `Operator`
+            The divergence operator used in composition.
+            Default:
+            Divergence(range=vector_field_jacobian.space, method='central')
+
+        Examples
+        --------
+        Identity deformation:
+
+        >>> import odl
+        >>> space = odl.uniform_discr(0, 1, 5)
+        >>> pspace = ProductSpace(space, space.ndim)
+        >>> vector_field = pspace.element(space.points().T)
+        >>> jacobian = space.zero()
+        >>> deformation = MassPreservingLinearDeformation(vector_field,
+        ...                                               jacobian)
+        """
+
+        assert isinstance(offsets.space, ProductSpace)
+        assert offsets.space.is_power_space
+        assert isinstance(offsets.space[0], DiscreteLp)
+        assert vector_field_jacobian in offsets.space[0]
+
+        ndim = offsets.space.size
+        space = offsets.space[0]
+
+        interp_points = space.points().T
+        for i in range(ndim):
+            interp_points[i] += offsets[i].ntuple.asarray()
+
+        self.interp_points = interp_points
+        self.offsets = offsets
+        self.vector_field_jacobian = vector_field_jacobian
+
+        #offsets.show('offsets')
+
+        if divergence is None:
+            self.divergence = Divergence(range=space, method='central')
+
+        Operator.__init__(self,
+                          domain=space,
+                          range=space,
+                          linear=True)
+
+    def _call(self, x):
+        result = x.interpolation(self.interp_points, bounds_check=False)
+        result = self.range.element(result)
+        result *= self.vector_field_jacobian
+        return result
+
+    def compose(self, linear_deformation):
+        assert isinstance(linear_deformation, LinearDeformation)
+
+        # Find the deformations for the output field
+        out_offsets = self.offsets.space.element()
+
+        for i in range(self.interp_points.shape[0]):
+            as_element = self.domain.element(self.interp_points[i])
+            interpolated = as_element.interpolation(
+                linear_deformation.interp_points, bounds_check=False)
+            out_offsets[i][:] = interpolated - self.domain.points().T[i]
+
+        # Find the jacobian
+        out_jacobian = self.domain.element(
+            self.vector_field_jacobian.interpolation(
+                linear_deformation.interp_points, bounds_check=False))
+
+        # Weighting
+        div = self.divergence(linear_deformation.offsets)
+        out_jacobian *= np.exp(div)
```

Comment:
Well this all breaks with large deformations, the divergence can go to infinity (this happens if your step is to large in the OIT code). Sadly the theory is all about "epsilon sized deformations".